# Career Track Capstone Project 1
## Data Cleaning Part I
### Data Source
The data set come from the [Yelp's Open Dataset](https://www.yelp.com/dataset). The documentation of the data set is [here](https://www.yelp.com/dataset/documentation/json).

### Import Libraries and data
At first, we import the data of "business". It include business location, attributes, and categories, etc.

In [1]:
import json
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import time
import datetime

filename = 'business.json'

js = [json.loads(line) for line in open(filename)]

business_data = pd.DataFrame(js)

### Checking data structure
We firstly check the names of the columns, and the brief information of the dataframe.

In [2]:
business_data.columns

Index(['address', 'attributes', 'business_id', 'categories', 'city', 'hours',
       'is_open', 'latitude', 'longitude', 'name', 'neighborhood',
       'postal_code', 'review_count', 'stars', 'state'],
      dtype='object')

In [3]:
business_data.shape

(174567, 15)

In [4]:
business_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 174567 entries, 0 to 174566
Data columns (total 15 columns):
address         174567 non-null object
attributes      174567 non-null object
business_id     174567 non-null object
categories      174567 non-null object
city            174567 non-null object
hours           174567 non-null object
is_open         174567 non-null int64
latitude        174566 non-null float64
longitude       174566 non-null float64
name            174567 non-null object
neighborhood    174567 non-null object
postal_code     174567 non-null object
review_count    174567 non-null int64
stars           174567 non-null float64
state           174567 non-null object
dtypes: float64(3), int64(2), object(10)
memory usage: 20.0+ MB


In [5]:
business_data.describe()

is_open       latitude      longitude   review_count  \
count  174567.000000  174566.000000  174566.000000  174567.000000   
mean        0.840376      38.627312     -92.679009      30.137059   
std         0.366258       5.389012      26.240079      98.208174   
min         0.000000     -36.086009    -142.466650       3.000000   
25%         1.000000      33.631550    -112.125879       4.000000   
50%         1.000000      36.144257     -89.410128       8.000000   
75%         1.000000      43.606181     -79.657609      23.000000   
max         1.000000      89.999314     115.086769    7361.000000   

               stars  
count  174567.000000  
mean        3.632196  
std         1.003739  
min         1.000000  
25%         3.000000  
50%         3.500000  
75%         4.500000  
max         5.000000

### Checking value frequency of each variable
Next, let's have a look about the value frequency of the 'state' varible.

In [6]:
business_data.state.value_counts(dropna=False).sort_values()

           1
B          1
MT         1
GA         1
MN         1
NE         1
WA         1
FL         1
KHL        1
PKN        1
AB         1
3          1
30         1
NLK        1
HU         1
TAM        1
DE         1
ZET        1
SL         1
STG        1
KY         1
CS         1
AL         1
FAL        1
WHT        1
VA         1
RCC        1
AK         1
CMA        2
AR         2
       ...  
BY         4
XGL        4
SCB        5
CA         5
VS         7
NI        10
01        10
ST        11
ESX       12
NY        18
C         28
WLN       38
ELN       47
FIF       85
CHE      143
NYK      152
HLD      179
MLN      208
SC       679
IL      1852
BW      3118
EDH     3795
WI      4754
QC      8169
PA     10109
OH     12609
NC     12956
ON     30208
NV     33086
AZ     52214
Name: state, Length: 68, dtype: int64

'BW' is an area in Germany. Here we just need states of USA. 
In addition, the number of business entities should be big enough to portrait a place.
So we choose states in USA which has more than 1000 business entities in our data set.

In [7]:
usa_states = business_data['state'].isin(['AZ','IL','NC','NV','OH','PA','WI'])

In [8]:
business_usa = business_data[usa_states].sort_values(by=['state'])
business_usa = business_usa.reset_index(drop=True)
business_usa.head()

address  \
0                4855 E Warner Rd, Ste B9   
1                                           
2                          4240 W Bell Rd   
3  17570 N 75th Ave, Ste 540, Room Ste 22   
4                    1904 E Washington St   

                                          attributes             business_id  \
0  {'AcceptsInsurance': True, 'ByAppointmentOnly'...  FYWN1wneV18bWNgQjJ2GNg   
1                                                 {}  IAPGit0LwvhR1fe07rK7bA   
2  {'BusinessAcceptsCreditCards': True, 'Restaura...  JfUt7soDtiGaBvseF0LCNA   
3  {'BusinessParking': {'garage': False, 'street'...  7FhjdyZ3jqD-YvF43oGgmg   
4               {'BusinessAcceptsCreditCards': True}  Vdr_Gp6uJnec3GETtjKJlw   

                                          categories       city  \
0  [Dentists, General Dentistry, Health & Medical...  Ahwatukee   
1                       [Landscaping, Home Services]    Phoenix   
2  [Home & Garden, Mattresses, Shopping, Furnitur...   Glendale   
3  [Hair Stylists, Hair Salons, Waxing, Hair Remo...   Glendale   
4  [Powder Coating, Furniture Reupholstery, Inter...    Phoenix   

                                               hours  is_open   latitude  \
0  {'Friday': '7:30-17:00', 'Tuesday': '7:30-17:0...        1  33.330690   
1                                                 {}        1  33.468804   
2  {'Monday': '10:00-20:00', 'Tuesday': '10:00-20...        1  33.640832   
3  {'Tuesday': '10:00-20:00', 'Friday': '10:00-18...        1  33.643939   
4  {'Monday': '8:00-16:00', 'Tuesday': '8:00-16:0...        1  33.448611   

    longitude                name neighborhood postal_code  review_count  \
0 -111.978599    Dental by Design                    85044            22   
1 -112.196673  Reparo Landscaping                    85035             3   
2 -112.150735      Mega Furniture                    85308            12   
3 -112.221370         Salon Haven                    85308             6   
4 -112.040803  Nu Look Revinyling                    85034             3   

   stars state  
0    4.0    AZ  
1    5.0    AZ  
2    2.5    AZ  
3    5.0    AZ  
4    4.0    AZ

Now we will check the values of other variables like 'address','business_id','city', etc., one by one. 

In [9]:
business_usa.address.value_counts(dropna=False).sort_index().head()

                                 5974
, 2190 East Williams Field Rd       1
, 3107 West Chandler Blvd           1
, 3200 S Las Vegas Blvd             1
, 3930 Las Vegas Blvd S             1
Name: address, dtype: int64

In [10]:
business_usa.business_id.value_counts(dropna=False).sort_index().head()

--7zmmkVg-IMGaXbuVd0SQ    1
--8LPVSo5i0Oo61X01sV9A    1
--9QQLMTbFzLJ_oT-ON3Xw    1
--9e1ONYQuAa-CB_Rrw7Tw    1
--DdmeR16TRb3LsjG0ejrQ    1
Name: business_id, dtype: int64

In [11]:
business_usa.city.value_counts(dropna=False).sort_index().head()

110 Las Vegas                   1
Ahwahtukee                      1
Ahwatukee                      16
Ahwatukee Foothills Village     1
Allegheny                       1
Name: city, dtype: int64

In [12]:
business_usa.is_open.value_counts(dropna=False).sort_index().head()

0     19858
1    107722
Name: is_open, dtype: int64

In [13]:
business_usa.latitude.value_counts(dropna=False).sort_values().head()

40.439371    1
33.306078    1
33.516018    1
33.418402    1
33.727981    1
Name: latitude, dtype: int64

In [14]:
business_usa.longitude.value_counts(dropna=False).sort_index().tail()

-75.062989     1
-74.858958     1
-74.150722     1
-71.780522     1
 115.086769    1
Name: longitude, dtype: int64

In [15]:
business_usa.name.value_counts(dropna=False).sort_index().head()

"T"s Hair Affair                       1
# 1 Nails                              1
#1 Cochran Buick GMC of Monroeville    1
#1 Cochran Buick GMC of Robinson       1
#1 Cochran Cadillac - Monroeville      2
Name: name, dtype: int64

In [16]:
business_usa.neighborhood.value_counts(dropna=False).sort_index().head()

                       89680
200th Street               2
Allentown                 16
Allied Dunn's Marsh       10
Anthem                  1208
Name: neighborhood, dtype: int64

In [17]:
business_usa.postal_code.value_counts(dropna=False).sort_index().head()
#Dealing with missing values

         410
02224      1
05452      1
08054      1
15003     36
Name: postal_code, dtype: int64

In [18]:
business_usa.review_count.value_counts(dropna=False).sort_index().head()

3    21748
4    11896
5     9075
6     7184
7     5960
Name: review_count, dtype: int64

In [19]:
business_usa.stars.value_counts(dropna=False).sort_index()

1.0     2925
1.5     3193
2.0     6921
2.5    11665
3.0    15734
3.5    21824
4.0    23041
4.5    18272
5.0    24005
Name: stars, dtype: int64

### Filling the missing values and correcting the variable with wrong values

From the information above, we find that the 'address', 'neighborhood' and 'postal_code' both have some missing values. We can use the 'latitude' and the 'longitude' to inquire related 'address', 'neighborhood' and 'postal_code' via the Google Geocoding API. 

The 'address' has 5974 missing values, the 'neighborhood' has 89680 missing values, and the 'postal_code' has 410 missing values.

Considering the [Google Geocoding API Usage Limits](https://developers.google.com/maps/documentation/geocoding/usage-limits), we choose to fill the missing values of the 'postal_code'. 

We also need to get the right values of the 'city', '110 Las Vegas' definitely is not a right city name. If we want to do statistics on the state level, and the city level, we need right cities' names.

We cannot correct all cities' names with iterating each observation. After filling the missing values of the 'postal_code', we can group observations by 'state', 'postal_code' and 'city', and get  one set of 'latitude' and 'longitude', representing the location coordinates of each city, to inquire the city name from Google Geocoding API. Thus the number of requests will below the 2500 limitation.

Here are two functions. The 'get_address' is to get 'address','postal_code', 'neighborhood' and 'city' with given 'latitude' and 'longitude'. Here API_KEY is set as 'None' as the free requesting. Sometime Google API will send back response as out of daily limit. We can wait for a while then do it again.

In [2]:
# Funtions Dealing with missing values for address,postal_code and neighborhood via lat and long
import requests

API_KEY = None
# Backoff time sets how many minutes to wait between google pings when your API limit is hit
BACKOFF_TIME = 30
OPT_LIST = ['address','postal_code','neighborhood','city']

def get_address(latitude,longitude,opt=None,api_key=None):
    result = {} 
    sensor = 'true'
    http_prefix = "https://maps.googleapis.com/maps/api/geocode/json?"
    params = "latlng={lat},{lon}&sensor={sen}".format(
        lat=latitude,
        lon=longitude,
        sen=sensor
    )
    url = "{http_prefix}{params}".format(http_prefix=http_prefix, params=params)
    if api_key is not None:
        url = url + "&key={}".format(api_key)
    while True:
        response = requests.get(url).json()
        if response['status'] == 'OK':
            #print("google response:",response)
            #extract_val = []
            result['res'] = 'OK'
            if opt == 'address' :
                result['content'] = response['results'][0]['formatted_address']
            if opt == 'city' :
                extract_val = [x['long_name'] for x in response['results'][0].get('address_components') 
                                if 'locality' in x.get('types')] 
                if len(extract_val)==0 :
                    extract_val = [x['long_name'] for x in response['results'][0].get('address_components') 
                                if 'neighborhood' in x.get('types')] 
                    if len(extract_val)==0 :
                        #result['res']= 'Extract nothing from response info'
                        result['content'] = ''
                        print(latitude,longitude)
                    else:
                        result['content'] = extract_val[0]
                else:
                    result['content'] = extract_val[0]
            if opt == 'postal_code' :
                extract_val = [x['long_name'] for x in response['results'][0].get('address_components') 
                                if 'postal_code' in x.get('types')]   
                if len(extract_val)==0 :
                    result['content'] = ''
                    print(latitude,longitude)
                else:
                    result['content'] = extract_val[0]
            if opt == 'neighborhood' :
                extract_val = [x['long_name'] for x in response['results'][0].get('address_components') 
                                if 'neighborhood' in x.get('types')]
                if len(extract_val)==0 :
                    result['content'] = ''
                    print(latitude,longitude)
                else:
                    result['content'] = extract_val[0]
        else :
            result['res']= response['status']
            result['content'] = None
        break
    return result

def reverse_geocoding(df_reversing,opt):
    result={}
    contents = []
    content_results = []
    if opt in OPT_LIST :
        for index,row in df_reversing.iterrows():
            
            gecd_reversed = False
            while gecd_reversed is not True:
               # Reverse the lat and long to address with google
                try:
                    reverse_result = get_address(row['latitude'],row['longitude'],opt=opt,api_key=API_KEY)
                except Exception as e:
                    gecd_reversed = True 
                if reverse_result['res'] == 'OK':
                    contents.append(reverse_result['content']) 
                    content_results.append(reverse_result['res'])
                    gecd_reversed = True 
                else:
                    print(reverse_result)                    
                    gecd_reversed = False
                    time.sleep(BACKOFF_TIME * 0.2)
                    
        df_reversing[opt] = pd.Series(contents).values
        result['res']='OK'
        result['content']=df_reversing[opt]
        result['content_result'] = content_results
    else :
        result['res']='Option Error'
        result['content']=''
        result['content_result']=[]
    return result

def get_city_coor(lat_lon_s):
    length = len(lat_lon_s)//2
    return lat_lon_s.values[length]

We firstly inquire missing postal codes. If we cannot get postal code from the response of Google API, we print out the set of latitude and longitude, and assign '' to the 'postal_code'. Later we will delete them from our dataset.

In [26]:
bool_post = business_usa['postal_code']==''
rev_df = business_usa[bool_post][['latitude','longitude']]
query_post = reverse_geocoding(rev_df,'postal_code')
if query_post['res'] == 'OK' :
    missed_post = query_post['content'] 

36.3997112534 -115.189260704


In [31]:
business_usa.update(missed_post)
business_usa['postal_code'] = business_usa['postal_code'].astype('str') 
business_usa = business_usa[business_usa['postal_code']!= '']

Then we will deal with the 'city' variable. If we cannot get city name from Google API, we will set it back to the old name. Because the 'latitude' and the 'longitude' are probably wrong. 

In [33]:
cities_coor = business_usa.groupby(['state','postal_code','city'])[['latitude','longitude']].agg(get_city_coor).reset_index()
cities_coor['city_y'] = cities_coor['city']
cities_dict = reverse_geocoding(cities_coor,'city')

33.4782016 -111.2392871
33.4091 -111.2181
41.3817911505 -81.9371392578
41.465503850000005 -81.2156177
41.464516383562504 -81.22236660261875
41.37497261666667 -81.93425878333333
39.320877215 -84.3674090132


In [34]:
unkown = {}
for index, row in cities_coor.iterrows():
    if row['city']=='':
        unkown[index] = row['city_y']

unkown_city = pd.Series(unkown,name='city')
cities_coor.update(unkown_city)
cities_coor['postal_code'] = cities_coor['postal_code'].astype('str') 

We use those new cities' names to update the 'city' column in our dataframe. 

In [36]:
city_dict = {}
for index, row in business_usa.iterrows():
    bl = (cities_coor['state']==row['state']) & (cities_coor['postal_code']==row['postal_code']) & (cities_coor['city_y']==row['city'])
    city = cities_coor[bl].city.unique()[0]
    #print("yelp city name:",row['city'], " google city name:",city)
    city_dict[index] = city
city_g = pd.Series(city_dict,name='city')
business_usa.update(city_g)


### Save data to a file for later use 
Now, we fill the missing values of the 'postal_code', and correct  the values of the 'city'. We can write the new business DataFrame to a CSV file for the use later. 

In [ ]:
business_usa.to_csv("{}_filled_postal-city_{}.csv".format('business_usa',datetime.datetime.now()))